pkg/kubectl/cmd/log.go

```diff
@@ -50,137 +49,157 @@ $ kubectl logs --tail=20 nginx
 	cmd.Flags().String("since-time", "", "Only return logs after a specific date (RFC3339). Defaults to all logs. Only one of since-time / since may be used.")
-	cmd.Flags().Duration("since", 0, "Only return logs newer than a relative duration like 5s, 2m, or 3h. Defaults to all logs. Only one of since-time / since may be used.")
-	cmd.Flags().StringVarP(&params.containerName, "container", "c", "", "Container name")
+	cmd.Flags().DurationVar(&o.SinceSeconds, "since", o.SinceSeconds, "Only return logs newer than a relative duration like 5s, 2m, or 3h. Defaults to all logs. Only one of since-time / since may be used.")
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", o.ContainerName, "Container name")
 	return cmd
 }
 
-// RunLog retrieves a pod log
-func RunLog(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, p *logParams) error {
-	if len(os.Args) > 1 && os.Args[1] == "log" {
-		printDeprecationWarning("logs", "log")
-	}
-
-	if len(args) == 0 {
+func (o *LogsOptions) Complete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string) error {
+	switch len(args) {
+	case 0:
 		return cmdutil.UsageError(cmd, "POD is required for log")
-	}
 
-	if len(args) > 2 {
-		return cmdutil.UsageError(cmd, "log POD [CONTAINER]")
-	}
+	case 1:
+		o.PodName = args[0]
+	case 2:
+		o.PodName = args[0]
+		o.ContainerName = args[1]
 
-	sinceSeconds := cmdutil.GetFlagDuration(cmd, "since")
-	sinceTime := cmdutil.GetFlagString(cmd, "since-time")
-	if len(sinceTime) > 0 && sinceSeconds > 0 {
-		return cmdutil.UsageError(cmd, "only one of --since, --since-time may be specified")
+	default:
+		return cmdutil.UsageError(cmd, "log POD [CONTAINER]")
 	}
 
-	namespace, _, err := f.DefaultNamespace()
+	var err error
+	o.PodNamespace, _, err = f.DefaultNamespace()
 	if err != nil {
 		return err
 	}
-	client, err := f.Client()
+	o.Client, err = f.Client()
 	if err != nil {
 		return err
 	}
 
-	podID := args[0]
+	sinceTime := cmdutil.GetFlagString(cmd, "since-time")
+	if len(sinceTime) > 0 {
+		t, err := api.ParseRFC3339(sinceTime, unversioned.Now)
+		if err != nil {
+			return err
+		}
+		o.SinceTime = &t
+	}
 
-	pod, err := client.Pods(namespace).Get(podID)
+	return nil
+}
+
+func (o *LogsOptions) Validate(cmd *cobra.Command) error {
+	if len(o.PodName) == 0 {
+		return cmdutil.UsageError(cmd, "POD must be specified")
+	}
+	if o.LimitBytes < 0 {
+		return cmdutil.UsageError(cmd, "--limit-bytes must be greater than or equal to zero")
+	}
+	if o.Tail < -1 {
+		return cmdutil.UsageError(cmd, "--tail must be greater than or equal to -1")
+	}
+	if o.SinceTime != nil && o.SinceSeconds > 0 {
+		return cmdutil.UsageError(cmd, "only one of --since, --since-time may be specified")
+	}
+
+	return nil
+}
+
+// RunLog retrieves a pod log
+func (o *LogsOptions) RunLog() error {
+	pod, err := o.Client.Pods(o.PodNamespace).Get(o.PodName)
```

Comment:
> Can we use the resource builder with pod resources defaulted?

I don't think I want to do it here.  In fact, I'm not certain its the right abstraction for what we want here.  In writing https://github.com/openshift/origin/pull/4997, I realized that we don't actually want to retrieve the referenced resource in all cases.  In some cases, we simply use it for referential lookups of the resources we're interested in.